src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Cache.cs

```diff
@@ -211,24 +217,40 @@ public JsonPropertyInfo GetProperty(ReadOnlySpan<byte> propertyName, ref ReadSta
                 }
             }
 
-            // No cached item was found. Try the main list which has all of the properties.
-
-            string stringPropertyName = JsonHelpers.Utf8GetString(propertyName);
-
+            // No cached item was found. Try the main dictionary which has all of the properties.
             Debug.Assert(PropertyCache != null);
-
-            if (!PropertyCache.TryGetValue(stringPropertyName, out info))
+            JsonPropertyInfo jsonPropertyInfo;
+            string stringPropertyName = JsonHelpers.Utf8GetString(propertyName);
+            if (PropertyCache.TryGetValue(stringPropertyName, out jsonPropertyInfo!))
             {
-                info = JsonPropertyInfo.s_missingProperty;
-            }
+                if (Options.PropertyNameCaseInsensitive)
+                {
+                    if (propertyName.SequenceEqual(jsonPropertyInfo!.NameAsUtf8))
+                    {
+                        Debug.Assert(key == GetKey(jsonPropertyInfo!.NameAsUtf8.AsSpan()));
```

Comment:
Then you should be able to remove all of these `!`s as well.

Reply:
Yep. Thanks